src/network/network_udp.cpp

```diff
@@ -33,19 +33,52 @@
 
 #include "../safeguards.h"
 
-/** Mutex for all out threaded udp resolution and such. */
-static std::mutex _network_udp_mutex;
-
 /** Session key to register ourselves to the master server */
 static uint64 _session_key = 0;
 
 static const std::chrono::minutes ADVERTISE_NORMAL_INTERVAL(15); ///< interval between advertising.
 static const std::chrono::seconds ADVERTISE_RETRY_INTERVAL(10); ///< re-advertise when no response after this amount of time.
 static const uint32 ADVERTISE_RETRY_TIMES     =              3; ///< give up re-advertising after this much failed retries
 
-NetworkUDPSocketHandler *_udp_client_socket = nullptr; ///< udp client socket
-NetworkUDPSocketHandler *_udp_server_socket = nullptr; ///< udp server socket
-NetworkUDPSocketHandler *_udp_master_socket = nullptr; ///< udp master socket
+static bool _network_udp_server;         ///< Is the UDP server started?
+static uint16 _network_udp_broadcast;    ///< Timeout for the UDP broadcasts.
+static uint8 _network_advertise_retries; ///< The number of advertisement retries we did.
+
+/** Some information about a socket, which exists before the actual socket has been created to provide locking and the likes. */
+struct UDPSocket {
+	const std::string name;                     ///< The name of the socket
+	std::mutex mutex;                           ///< Mutex for everything that (indirectly) touches the sockets within the handler.
+	NetworkUDPSocketHandler *socket;            ///< The actual socket, which may be nullptr when not initialized yet.
+	std::atomic<int> receive_iterations_locked; ///< The number of receive iterations the mutex was locked.
+
+	UDPSocket(const std::string &name_) : name(name_), socket(nullptr) {}
+
+	void Close()
+	{
+		std::lock_guard<std::mutex> lock(mutex);
+		socket->Close();
+		delete socket;
+		socket = nullptr;
+	}
+
+	void ReceivePackets()
+	{
+		std::unique_lock<std::mutex> lock(mutex, std::defer_lock);
+		if (!lock.try_lock()) {
+			if (receive_iterations_locked++ % 32 == 0) {
```

Comment:
shouldn't this be `++receive_iterations_locked` ? As now the first iteration always returns 0? (honest question, these constructs confuse the fuck out of me)

Reply:
Yes, you're right. Now you also get a message for the first iteration... which might be a bit too spammy.